python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -0,0 +1,172 @@
+from typing import Dict, Any, Iterator, Callable, List, Tuple
+import ray
+from ray.data.block import Block, BlockAccessor, BlockMetadata, BlockExecStats
+from ray.data.context import DEFAULT_SCHEDULING_STRATEGY, DatasetContext
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data._internal.execution.operators.map_task_submitter import MapTaskSubmitter
+from ray.types import ObjectRef
+
+
+class ActorPoolSubmitter(MapTaskSubmitter):
+    """A task submitter for MapOperator that uses a Ray actor pool."""
+
+    def __init__(self, pool_size: int, ray_remote_args: Dict[str, Any]):
+        """Create an ActorPoolSubmitter instance.
+
+        Args:
+            pool_size: The size of the actor pool.
+            ray_remote_args: Remote arguments for the Ray actors to be created.
+        """
+        self._pool_size = pool_size
+        self._ray_remote_args = ray_remote_args
+        # A map from task output futures to the actors on which they are running.
+        self._active_actors: Dict[ObjectRef[Block], ray.actor.ActorHandle] = {}
+        # The actor pool on which we are running map tasks.
+        self._actor_pool = ActorPool()
+
+    def start(self):
+        # Check that actor pool hasn't yet been initialized.
+        assert self._actor_pool.num_actors == 0
+        # Create the actor workers and add them to the pool.
+        ray_remote_args = self._apply_default_remote_args(self._ray_remote_args)
+        cls_ = ray.remote(**ray_remote_args)(MapWorker)
+        for _ in range(self._pool_size):
+            self._actor_pool.add_actor(cls_.remote())
```

Comment:
@ericl Since the actor pool is static and I'm trying to keep this first-pass pool simple, we currently don't wait until the actor is ready to add it to the pool. I think that this could lead to liveness issues, e.g. if a map task is submitted to an actor that will never start due to resource constraints, while there are other (running) actors available.

There seem to be a few possible solutions:
1. Incorporate the `actor.ready.remote()` futures into the executor control loops.
2. Add a control loop to the execution state or actor pool submitter, similar to the `ActorPoolStrategy` control loop.
3. Truncate the pool size to the available resources in the cluster to get a soft assurance of liveness.
4. When picking an actor, choose the actor with that smallest task queue that is also running.
5. At `ActorPool.pick_actor()` time (i.e. task submission time), block until an actor is running.

I think that a combination of (4) and (5) could work well for this PR, with (3) being a nice-to-have, and (1) or (2) being the eventual solution needed for the streaming executor. What do you think?

Reply:
> Incorporate the actor.ready.remote() futures into the executor control loops.

Yep, that's what I was thinking we could do (return this as party of work_refs from the operator). We could leave this as a TODO as this PR is already quite large, up to you.

Reply:
> When picking an actor, choose the actor with the smallest task queue that is also running.

You'd still need to wait on a future to check if it's running right? I think that circles back to implementing (1) properly.

Reply:
@ericl Yep, and agreed that the cleanest solution would involve (1), but one idea for implementing (4) + (5) now (without executor control loop integration) would be doing a `ray.wait()` check at pick time:
```python
def add_actor(self, actor: ray.actor.ActorHandle):
    self._pending_actors[actor.ready.remote()] = actor

def pick_actor(self) -> ray.actor.ActorHandle:
    if self._pending_actors:
        # At least one actor still pending, block until at least one actor is ready,
        # and check if any pending actors are newly ready.
        it not self._tasks_in_flight:
            # No actors ready, block until at least one actor is ready.
            self._wait_for_ready_actors(block_for_one=True)
        # At least one actor ready, quickly check if more actors are now ready.
        self._wait_for_ready_actors(block_for_one=False)

    actor = min(
         self._tasks_in_flight.keys(), key=lambda actor: self._tasks_in_flight[actor]
    )
    self._tasks_in_flight[actor] += 1
    return actor

def _wait_for_ready_actors(self, block_for_one: bool):
    ready, _ = ray.wait(
        list(self._pending_actors.keys()),
        num_returns=1 if block_for_one else len(self._pending_actors),
        fetch_local=False,
        timeout=None if block_for_one else 0)
    for future in ready:
        actor = self._pending_actor.pop(future)
        self._tasks_in_flight[actor] = 0
```

If we're ok with letting this liveness issue slip by in this PR, we can wait to properly implement (1) in a future PR. Would you prefer to wait, or have something like the above implemented in this PR?

Reply:
Let's just do it properly in another PR--- I think it's a pretty small change and we don't need to worry about it in the mean-time.

Reply:
Sure, sounds good!